internal/consensus/state.go

```diff
@@ -2264,54 +2311,59 @@ func (cs *State) addVote(vote *types.Vote, peerID p2p.ID) (added bool, err error
 		"extLen", len(vote.Extension),
 		"extSigLen", len(vote.ExtensionSignature),
 	)
+}
 
-	if vote.Height < cs.Height || (vote.Height == cs.Height && vote.Round < cs.Round) {
-		cs.metrics.MarkLateVote(vote.Type)
-	}
+// isLateVote checks if a vote is a late vote.
+func (cs *State) isLateVote(vote *types.Vote) bool {
+	return vote.Height < cs.Height || (vote.Height == cs.Height && vote.Round < cs.Round)
+}
 
-	// A precommit for the previous height?
-	// These come in while we wait timeoutCommit
-	if vote.Height+1 == cs.Height && vote.Type == types.PrecommitType {
-		if cs.Step != cstypes.RoundStepNewHeight {
-			// Late precommit at prior height is ignored
-			cs.Logger.Debug("precommit vote came in after commit timeout and has been ignored", "vote", vote)
-			return added, err
-		}
+// isPrecommitForPreviousHeight checks if a vote is a precommit for the previous height.
+func (cs *State) isPrecommitForPreviousHeight(vote *types.Vote) bool {
+	return vote.Height+1 == cs.Height && vote.Type == types.PrecommitType
+}
 
-		added, err = cs.LastCommit.AddVote(vote)
-		if !added {
-			// If the vote wasn't added but there's no error, its a duplicate vote
-			if err == nil {
-				cs.metrics.DuplicateVote.Add(1)
-			}
-			return added, err
-		}
+// handlePrecommitForPreviousHeight handles a precommit vote for the previous height.
+func (cs *State) handlePrecommitForPreviousHeight(vote *types.Vote) (bool, error) {
+	if cs.Step != cstypes.RoundStepNewHeight {
```

Comment:
I know this was in the original code, but I don't see much point in this test with debug-only purposes.

We only add votes to `cs.LastCommit` when it is a precommit from the previous height **and** we are in the new_height round step.